File: Problem_Solvings/백준/백트래킹/9663/9663.py
def promising(queens,new_column):
    current_row=len(queens)
    for row in range(current_row):
        if queens[row]==new_column or current_row-row==abs(new_column-queens[row]):
            return False
    return True

def n_queens(num,queens,count):
    if len(queens)==num:
        count[0]+=1

    for i in range(num):
        if promising(queens,i):
           n_queens(num,queens+[i],count)


def solution():
    num=int(input())
    count=[0]
    n_queens(num,[],count)
    print(count[0])
if __name__ == "__main__":
    solution()

File: Problem_Solvings/백준/백트래킹/9663/9663.c
#define _CRT_SECURE_NO_WARNINGS
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#define TRUE 1
#define FALSE 0

int count = 0;
int *rows;
int promising(int current_row, int next_column) {
	int i = 0;
	int next_row = current_row;
		for (i = 0; i < current_row; i++) {
			if (rows[i] == next_column || next_row - i == abs(rows[i] - next_column))
				return FALSE;
		}
	return TRUE;
}

void nqueens(int num, int current_row) {
	int i = 0;
		if (current_row == num) {
			count += 1;
			return;
		}
	for (i = 0; i < num; i++) {
		if (promising(current_row, i)) {
			rows[current_row] = i;
			nqueens(num, current_row + 1);
			rows[current_row] = 0;
		}
	}
}
int main(void) {
	int i, num;
	scanf("%d", &num);
	rows = (int*)malloc(sizeof(int)*(num));
	for (i = 0; i < num; i++)
		rows[i] = 0;
	nqueens(num,0);
	printf("%d", count);
	free(rows);
}